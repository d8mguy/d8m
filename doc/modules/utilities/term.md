## Term

Module Term provides a `Term` type suitable for parser outputs, and which is used by the lexparse module.

#### Term:type

    val Term = extend tuple(kind: string, arg: ortype(terms: list(Term), strg: string), loc:integer) with {
        method stringify : \() -> string
        method asString = \() -> string
        method asTerms = \() -> list(Term)
        method child0 = \() -> Term
        method child1 = \() -> Term
        method child2 = \() -> Term
        method childN = \(n:integer) -> Term
        method asLeaf = \() -> string       // get the value of a leaf node
        method hasKind = \(wanted:string) -> boolean {...}
        method hasMatch = \(matchfn: \(Term)->boolean) -> boolean {...}
    }

The `asString` and `asTerms` methods interpret `self.arg` as a specific type. These methods are normally applied based on the value of `kind`; if lexparse creates the Term, `kind` is determined by the production being reduced. The four `child` methods interpret self.arg as a list(Term) and return the child Term at the indicated index in the list; `asLeaf` treats child 0 as a string. The `hasKind` and `hasMatch` methods traverse self to determine whether a given kind occurs (for `hasKind`) or the `matchfn` matches on any child element of self.

Access to the Term's `loc` is intended to be accessed directly through the attribute, and is for recording of the location in a parsed string where a Term is found. Since Term isn't provided with a lithook method you must provide a `loc` in the litform creating each Term.
