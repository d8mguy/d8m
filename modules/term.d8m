
import "strings"

export Term

// A flexible version of Term and some methods to use it. Minimalist: just three attributes. The strg form of arg
// is for "leaf" nodes; the terms form is for internal (non-leaf) nodes. The loc slot is optional but can help
// convey info about where (presumably in a text file) the term comes from. Lexparse produces Terms.
// We define some helper methods like asString, asTerms, childN etc. Also a couple of recursive helpers: hasKind
// is a predicate on existence of a matching kind value anywhere in a Term; hasMatch looks for fn-based matches.
val Term: type = extend tuple(kind: string, arg: ortype(terms: list(Term), strg: string), loc:integer) where {
    // Note a subtlety: this is recursive so the rettype must be explicit
    method stringify = \() -> string {
        val args = (tag(arg) == :strg ? arg : strings.join(arg.{stringify(this)}, ";"))
        "[#{kind}: #{args}]"
    }
    method asString = \() { always(x~self.arg, tag(x) == :strg) }
    method asTerms = \() { always(x~self.arg, tag(x) == :terms) }
    method child0 = \() { asTerms(self)[0] }
    method child1 = \() { asTerms(self)[1] }
    method child2 = \() { asTerms(self)[2] }
    method childN = \(n:integer) { asTerms(self)[n] }
    method asLeaf = \() { asString(self.child0) }       // get the value of a leaf node
    method hasKind = \(wanted:string) -> boolean {
        if(tag(arg) == :strg) kind == wanted else any(arg.{ this.hasKind(wanted) })
    }
    method hasMatch = \(matchfn: \(Term)->boolean) -> boolean {
        if(matchfn(self)) true
        else if(tag(arg) == :strg) false
        else any(arg.{ this.hasMatch(matchfn)})
    }
}
