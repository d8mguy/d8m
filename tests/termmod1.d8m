// termmod1. Check that adding an item to a variant list works.
// See notes around 08/17/22 and 09/09/22. D8m semantics says that the pushb to fcptArgs should affect trm.arg since
// they're the same entity. This can work by taking pointers but you can't take pointers to an interface type in go.
// A hack in pass1 of cleanup causes fcptArgs to be substd by its binding so this test works. But it's not a completely
// correct implementation.
import "term" melted
var trm1 = [Term: "aaa", [[Term: "bbb", "ccc", 0]], 0]
val foo = \mod(trm: Term) {
    var fcptArgs = asTerms(trm)
    fcptArgs.pushb([Term: "children", "the", 0])
    trm
}
foo(trm1)
