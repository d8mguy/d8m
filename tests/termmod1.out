[aaa: [bbb: ccc];[children: the]]
